### PR TITLE
Minor refactor MetricReader and AggregationTemporality

### LINKED ISSUE #2355: Clean up Metric Reader/Exporter
* [x] https://github.com/open-telemetry/opentelemetry-dotnet/pull/2342#discussion_r708810437
* [x] https://github.com/open-telemetry/opentelemetry-dotnet/pull/2342#discussion_r708826103
* [x] https://github.com/open-telemetry/opentelemetry-dotnet/pull/2353#discussion_r708815138

### PR DESCRIPTION
**Pull request by reyang**
Trying to address the review comments / suggestions from @CodeBlanch.
Fixes #2355.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs [hunk 1]
```diff
@@ -21,7 +21,6 @@ namespace OpenTelemetry.Metrics
     public class BaseExportingMetricReader : MetricReader
     {
         protected readonly BaseExporter<Metric> exporter;
-        protected bool disposed;
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs:24]: Follow up on https://github.com/open-telemetry/opentelemetry-dotnet/pull/2342#discussion_r708810437.

I've moved `disposed` to the base class and made it private.
I guess I'll leave `exporter` as "protected readonly", this makes the code consistent with https://github.com/open-telemetry/opentelemetry-dotnet/blob/34d8332a809b6258f481fdd07deea1b52b5b080d/src/OpenTelemetry/BaseExportProcessor.cs#L29.

File: src/OpenTelemetry/Metrics/MetricReader.cs [hunk 2]
```diff
@@ -20,8 +20,10 @@ namespace OpenTelemetry.Metrics
 {
     public abstract class MetricReader : IDisposable
     {
-        private AggregationTemporality preferredAggregationTemporality = AggregationTemporality.Both;
-        private AggregationTemporality supportedAggregationTemporality = AggregationTemporality.Both;
+        private const AggregationTemporality CumulativeAndDelta = AggregationTemporality.Cumulative | AggregationTemporality.Delta;
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/MetricReader.cs:23]: Follow up on https://github.com/open-telemetry/opentelemetry-dotnet/pull/2342#discussion_r708826103.

I tried Both/Neither/All/None and **none** of them seems logical / good to me. So I stepped back and removed them, and use `Cumulative | Delta`.

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs [hunk 3]
```diff
@@ -30,15 +30,5 @@ public enum AggregationTemporality : byte
         /// Delta.
         /// </summary>
         Delta = 0b10,
-
```
Comment (reyang) [inline src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: Follow up on https://github.com/open-telemetry/opentelemetry-dotnet/pull/2353#discussion_r709361948.

I struggled between `AggregationTemporality` and `AggregationTemporalityModes`, it seems `AggregationTemporality` is a well established term in OpenTelemetry (IMHO), and `AggregationTemporalityModes` just looks ... weird to me.

@CodeBlanch @cijothomas @alanwest do you feel strong about this?

Comment (cijothomas) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: `AggregationTemporality` sounds good.. Yes  "modes" is bit weird

Comment (reyang) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: Yeah, I think `ExportModes` looks nice because `Export` is a noun.
I also struggled with `AggregationModes` for few seconds and concluded it was a bad idea.

File: src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs:24]: I don't have an issue with this just want to share the CA warning doc which has a bit of the reasoning behind not exposing fields: https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1051

TL;DR: The primary use of a field should be as an implementation detail.

File: src/OpenTelemetry/Metrics/MetricReader.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricReader.cs:23]: If it is going to be a common thing used a lot of places we could add on the enum like...

```csharp
    [Flags]
    public enum AggregationTemporality : byte
    {
        /// <summary>
        /// Cumulative.
        /// </summary>
        Cumulative = 0b1,

        /// <summary>
        /// Delta.
        /// </summary>
        Delta = 0b10,

        CumulativeAndDelta = Cumulative | Delta, 
    }
```

Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricReader.cs:23]: Given `CumulativeAndDelta = Cumulative BITWISE OR Delta` is weird, I think we can use the current approach (a private const) for now, introducing `AggregationTemporality.CumulativeAndDelta` in the future is still possible given it is not a breaking change.

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: I'm OK with keeping `AggregationTemporality`! I don't have an issue with `AggregationModes.Delta` either though. To me, being much further away from the metrics discussions, `AggregationModes.Delta` is immediately clear at a glance as to its purpose. "AggregationTemporality" I would need to google "temporality" to understand it. Here's what Google gave me: _the state of existing within or having some relationship with time_ 🧙‍♀️ 

But we could also clarify with XML comments.

**Review by CodeBlanch — APPROVED** [review]
I dropped some comments but nothing blocking.

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: Trying to be a bit creative here, how about:
* `AggregationModes.CumulativeTemporality`
* `AggregationModes.DeltaTemporality`

File: src/OpenTelemetry/Metrics/MetricReader.cs [hunk 4]
```diff
@@ -59,8 +61,12 @@ public virtual void OnCollect(Batch<Metric> metrics)
...
         public void Dispose()
         {
-            this.Dispose(true);
-            GC.SuppressFinalize(this);
+            if (!this.disposed)
+            {
+                this.Dispose(true);
+                GC.SuppressFinalize(this);
+                this.disposed = true;
+            }
```
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricReader.cs:69]: I think we don't have native resource so Finalizer is not needed (and having it will cause extra overhead)?

Or do you think we want to allow derived class to have native resource, and prevent folks from leaking them if they don't dispose the instances properly?

[the remark this answers is not in the document]
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricReader.cs:69]: > Or do you think we want to allow derived class to have native resource, and prevent folks from leaking them if they don't dispose the instances properly?

Exactly! We can't really control what users will do with it. Only time I think it is safe to leave out finalizer is if class is `internal` or `sealed` and not using any non-managed resources. I don't think there is a major perf hit for users correctly calling `Dispose` do to the `GC.SuppressFinalize(this);` call, but I haven't measured it.

[the remark this answers is not in the document]
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricReader.cs:69]: Just looking at the [docs](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose), I think my initial comment was messed up! Sorry. Looks like `MetricReader` should just be...

```csharp
public abstract class MetricReader : IDisposable
{
    public void Dispose()
    {
        this.Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool isDisposing)
    {
    }
}
```

Derived classes would be...

```csharp
public class MyMetricReader : MetricReader 
{
    private readonly SemaphoreSlim semaphore = new();
    private bool disposed;

    protected override void Dispose(bool isDisposing)
    {
        if (this.disposed)
        {
            return;
        }

        if (disposing)
        {
            this.semaphore.Dispose();
        }

        this.disposed = true;

        base.Dispose(disposing);
    }
}

public class MyMetricReaderWithNative : MetricReader 
{
    private IntPtr handle;
    private bool disposed;

    ~MyMetricReaderWithNative() => this.Dispose(false);

    protected override void Dispose(bool isDisposing)
    {
        if (this.disposed)
        {
            return;
        }

        // todo: clean up handle regardless of isDisposing

        this.disposed = true;

        base.Dispose(disposing);
    }
}
```

[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricReader.cs:69]: Related to #1024.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs [hunk 5]
```diff
@@ -32,7 +32,7 @@ namespace OpenTelemetry.Exporter
     /// Exporter consuming <see cref="Metric"/> and exporting the data using
     /// the OpenTelemetry protocol (OTLP).
     /// </summary>
-    [AggregationTemporality(AggregationTemporality.Both, AggregationTemporality.Cumulative)]
+    [AggregationTemporality(AggregationTemporality.Cumulative | AggregationTemporality.Delta, AggregationTemporality.Cumulative)]
```
Comment (alanwest) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs:35]: Not related to this PR, but I'm not following why the OTLP exporter is marked to prefer `Cumulative`. Seems to me the OTLP exporter should effectively not have a preference.

For a given service, is it currently not possible for some metrics to have a cumulative aggregation and others to have a delta aggregation?

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs (hunk 3 above)
Comment (alanwest) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: > Here's what Google gave me: _the state of existing within or having some relationship with time 🧙‍♀️_

😆 

Completely agree that the first time hearing "aggregation temporality" is definitely a "what did you just say to me?" moment... that said, to @reyang's point, it has become an established term in OpenTelemetry, so I lean more towards keeping `AggregationTemporality` in the name.

I'm a bit skeptical that `ExplicitReset`/`ImplicitReset` would ever be a thing, but I am interested if there's been any discussions about this possibility.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs (hunk 5 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs:35]: Ok, sorry, I _think_ this is making sense to me now. I'm still trying to keep all these things in my head. I continue going back to a time when there was discussion about being able to define aggregation temporality at the level of an individual instrument or view. This is not the case.

Aggregation temporality is essentially defined at the level of a pipeline - so like `MetricReader` -> `MetricExporter`.

Comment (reyang) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs:35]: Let me put some concrete examples to help us align the understanding:

Prometheus - only supports Cumulative (implied that Cumulative is also the preference):
* The user doesn't have to specify temporality during the SDK configuration, everything will be exported as cumulative. In case the measurement being reported is Delta, there will be Delta->Cumulative conversion.
* If the user specified Cumulative explicitly, it won't harm
* If the user specified Delta explicitly, they will get some error telling them that the setup is wrong.

OTLP - supports both Cumulative and Delta (and Cumulative is the preference):
* The user doesn't have to specify temporality during the SDK configuration, everything will be exported as Cumulative. In case the measurement being reported is Delta, there will be Delta->Cumulative conversion.
* If the user specified Cumulative explicitly, it won't harm
* If the user specified Delta explicitly, there are two cases:
  * If the value being reported is Delta (e.g. Counter), it'll be reported as Delta, no conversion required
  * If the value being reported is Cumulative (e.g. ObservableCounter), there will be Cumulative->Delta conversion.

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs (hunk 3 above)
Comment (cijothomas) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]:
```
AggregationModes.CumulativeTemporality
AggregationModes.DeltaTemporality
```

sounds good to me.

File: src/OpenTelemetry/Metrics/MetricReader.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (reyang) [reply src/OpenTelemetry/Metrics/MetricReader.cs]: Done.

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: We have 3 votes for `AggregationModes` and @alanwest prefers `AggregationTemporality`.
I will keep `AggregationTemporality` in this PR, once we merged it, I'll create a separate PR to rename it to `AggregationModes` and we can discuss more there.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs (hunk 5 above)
Comment (reyang) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs:35]: @alanwest FYI https://github.com/open-telemetry/opentelemetry-specification/issues/731

**Review by cijothomas — APPROVED** [review]

File: src/OpenTelemetry/Metrics/AggregationTemporality.cs (hunk 3 above)
Comment (reyang) [reply src/OpenTelemetry/Metrics/AggregationTemporality.cs:33]: Tracked by #2371.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs (hunk 5 above)
Comment (alanwest) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/OtlpMetricsExporter.cs:35]: Thank you for this context! Defaulting to cumulative had seemed somewhat arbitrary to me, but Prometheus + load-balanced collectors = a bad time if delta. Prometheus seems to make everything trickier 😄.
